cli/pcluster/utils.py

```diff
@@ -329,6 +329,87 @@ def get_supported_compute_instance_types(scheduler):
+    """
+    Return a tuple of availability zones that have the instance_types.
+
+    This function build above get_supported_az_for_multi_instance_types,
+    but simplify the input to 1 instance type and result to a list
+
+    :param instance_type: the instance type for which the supporting AZs.
+    :return: a tuple of the supporting AZs
+    """
+    return get_supported_az_for_multi_instance_types([instance_type])[instance_type]
+
+
+def get_supported_az_for_multi_instance_types(instance_types):
+    """
+    Return a dict of instance types to list of availability zones that have each of the instance_types.
+
+    :param instance_types: the list of instance types for which the supporting AZs.
+    :return: a dicts. keys are strings of instance type, values are the tuples of the supporting AZs
+
+    Example:
+    If instance_types is:
+    ["t2.micro", "t2.large"]
+    Result can be:
+    {
+        "t2.micro": (us-east-1a, us-east-1b),
+        "t2.large": (us-east-1a, us-east-1b)
+    }
+    """
+    # Most of the fancy code in the func is to try using only one API call for all instance types
+    if not hasattr(get_supported_az_for_multi_instance_types, "cache"):
+        get_supported_az_for_multi_instance_types.cache = {}
+    cache = get_supported_az_for_multi_instance_types.cache
+    missing_instance_types = []
+    result = {}
+    for instance_type in instance_types:
+        if instance_type in cache:
+            result[instance_type] = cache[instance_type]
+        else:
+            missing_instance_types.append(instance_type)
+    if missing_instance_types:
+        ec2_client = boto3.client("ec2")
+        paginator = ec2_client.get_paginator("describe_instance_type_offerings")
+        page_iterator = paginator.paginate(
+            LocationType="availability-zone", Filters=[{"Name": "instance-type", "Values": missing_instance_types}]
+        )
+        offerings = []
+        for page in page_iterator:
+            offerings.extend(page["InstanceTypeOfferings"])
+        for instance_type in missing_instance_types:
+            cache[instance_type] = tuple(
+                offering["Location"] for offering in offerings if offering["InstanceType"] == instance_type
+            )
+            result[instance_type] = cache[instance_type]
+    return result
+
+
+def get_availability_zone_of_subnet(subnet_id):
+    """
+    Return the availability zone of the subnet.
+
+    :param subnet_id: the id of the subnet.
+    :return: a strings of availability zone name
+    """
+    if not hasattr(get_availability_zone_of_subnet, "cache"):
+        get_availability_zone_of_subnet.cache = {}
+    cache = get_availability_zone_of_subnet.cache
+    if subnet_id not in cache:
+        try:
+            cache[subnet_id] = (
+                boto3.client("ec2").describe_subnets(SubnetIds=[subnet_id]).get("Subnets")[0].get("AvailabilityZone")
+            )
+        except ClientError as e:
```

Comment:
You are catching the exception here, so let's think about what is the behavior we want here.

Do we want the program to exit or continue? In either case, should we handle ClientError consistently in `get_supported_az_for_multi_instance_types` as well?

Reply:
Good catch. The code was adapted from the previous code. I will think more about the catch statement.